demos/demo-utils/src/logging.rs

```diff
@@ -42,39 +45,47 @@ impl<T: Timer, L: fmt::Write> fmt::Write for StampedLogger<T, L> {
 /// that we never block or drop data.
 pub struct BbqLogger {
     p: Producer,
+    data_lost: bool,
 }
 
 impl BbqLogger {
     pub fn new(p: Producer) -> Self {
-        Self { p }
+        Self {
+            p,
+            data_lost: false
+        }
     }
 }
 
 impl fmt::Write for BbqLogger {
     fn write_str(&mut self, s: &str) -> fmt::Result {
+        let data_lost_msg_bytes: usize = if self.data_lost {
+            DATA_LOST_MSG.as_bytes().len()
+        }
+        else {
+            0
+        };
```

Comment:
I think it would be easier to follow the code if you just did the `data_lost` logic before all of the regular logging logic. That way we wouldn't need to add the lengths together below. So the new logic would be:

```
if self.data_lost {
    // try to print DATA_LOST_MSG
    // if that succeeds, set `self.data_lost = false;` 
    // if there's not enough space for it, `return Ok(())`
}

// (old printing logic)
// not enough space => set `self.data_lost = true;`
```

Reply:
The problem with that approach is that we would get multiple DATA_LOST_MSG's printed repeatedly until there was enough space for the logging message to be sent, and at the same time the DATA_LOST_MSG would be starving the ring buffer preventing the logging message to enter. That's why I needed to account for both lengths together before sending anything, it either can send the data lost message plus the logging line, or do nothing and wait for the next check. I'll try to refactor it to be more clear.

Reply:
Ah, that makes sense. Thanks!